### PR TITLE
Feat(structN2V): allow cross patterns for structN2V

### PR DESCRIPTION
**Pull request by jdeschamps**
## Disclaimer

- [ ] I am an AI agent.
- [x] I have used AI and I thoroughly reviewed every line.
- [ ] I have not used AI extensively.

## Description

> [!NOTE]  
> **tldr**: Extends the structN2V masks to symmetric cross patterns. 

This PR introduces cross structN2V masks, and refactors the code-base to enable future arbitrary mask.

### Background - why do we need this PR?

Some users have requested symmetrical cross structN2V patterns, which can currently only be 1D.

### Overview - what changed?

The choices for structN2V masks have been extended to allow for `cross` pattern, the pixel manipulation functions have been updated to accommodate such a 2D pattern.

### Implementation - how did you implement the changes?

I did a small refactoring of the pixel manipulation functions to ensure that there is a single structN2V mask creation function, extended that function for 2D cross, and extended `SupportedStructAxis` and `StructN2VParameters`.

### New features or files

- Pixel manipulation: added `_build_struct_pattern` called in both the application of the mask and the median calculation, which create the structN2V mask, modified other functions to accommodate the new source of mask truth
- StructN2VParameters: added a constructor to convert axis key to index
     

### Modified features or files

- N2VManipulateConfig: added "cross"
- SupportedStructAxis: added "cross"
- Removed `_torch` from function names

### Removed features or files

- Removed old tests

## How has this been tested?

Added new tests and ran a notebook for mask creation.

## Related Issues

- Resolves #

## Breaking changes

No breaking change.

## Additional Notes and Examples

cross
horizontal
vertical

---

**Please ensure your PR meets the following requirements:**

- [x] Code builds and passes tests locally, including doctests
- [x] New tests have been added (for bug fixes/features)
- [x] Documentation has been updated
- [x] Pre-commit passes

### CODE REVIEW

File: src/careamics/lightning/modules/n2v_utils/struct_mask_parameters.py [hunk 1]
```diff
@@ -3,18 +3,63 @@
...
 @dataclass
 class StructMaskParameters:
     """Parameters of structN2V masks.
 
     Attributes
     ----------
-    axis : Literal[0, 1]
-        Axis along which to apply the mask, horizontal (0) or vertical (1).
+    axis : int
+        Axis along which to apply the mask, horizontal (0), vertical (1) or cross (2).
```
Comment (melisande-c) [inline src/careamics/lightning/modules/n2v_utils/struct_mask_parameters.py]: Having this be ints is now a bit confusing, since it originally is meant to be the axis on which the mask is applied and the value 2 implies the Z axis.

It should be either simply `"horizontal", "vertical", "cross"`, `SupportedStructAxis` or maybe `0, 1, (0, 1)` (a tuple of 0 and 1 to indicate both.)

File: src/careamics/config/support/supported_struct_axis.py [hunk 2]
```diff
@@ -12,10 +12,13 @@ class SupportedStructAxis(StrEnum):
         Horizontal axis.
     VERTICAL : str
         Vertical axis.
+    Cross : str
```
Comment (melisande-c) [inline src/careamics/config/support/supported_struct_axis.py]:
```suggestion
    CROSS : str
```

File: src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py [hunk 3]
```diff
@@ -333,18 +373,21 @@ def median_manipulate(
...
     return output_batch, mask
 
 
-def _create_center_pixel_mask(
+# -- Median pixel manipulation utilities
+
+
+def _create_neg_center_pixel_mask(
     ndims: int, subpatch_size: int, device: torch.device
 ) -> torch.Tensor:
```
Comment (melisande-c) [inline src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py]: In my mind the mask is not negative 🤷‍♀️

It is true where pixels should be selected for the median.

File: src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py [hunk 4]
```diff
@@ -38,37 +73,41 @@ def _apply_struct_mask_torch(
...
-    # Displacements from center
-    displacements = torch.stack(torch.where(mask == 1)) - center.unsqueeze(1)
-
-    # Combine all coords (ndim, npts) with all displacements (ncoords, ndim)
-    mix = displacements.T.unsqueeze(-1) + coords.T.unsqueeze(0)
-    mix = mix.permute([1, 0, 2]).reshape([mask.ndim, -1]).T
+    # build 2D (span x span) displacement mask: 1 where pixels should be replaced
+    disp_mask_2d = _build_struct_pattern(
+        struct_params.span, struct_params.axis, patch.device
+    )
```
Comment (melisande-c) [inline src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py:82]: I think `_apply_struct_mask` was originally following closely the original implementation, variable names and all, but it is a little hard to follow. https://github.com/juglab/n2v/blob/c61b3b47c5cacc533f7e0e3db7cefab139d7b62a/n2v/internals/N2V_DataWrapper.py#L98-L116 

But I guess it's not that important, we can leave for another time.

File: src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py [hunk 5]
```diff
@@ -400,16 +442,18 @@ def _create_struct_mask(
...
     struct_mask = torch.ones((subpatch_size,) * ndims, dtype=torch.bool, device=device)
-    # indexes the center unless it is the axis on which the struct mask spans
-    struct_slice = (
-        center_idx if d != span_axis else slice(span_start, span_end)
-        for d in range(ndims)
-    )
-    struct_mask[*struct_slice] = False
+    # leading dims (e.g. Z for 3D) are fixed at the center index
+    leading = (center_idx,) * (ndims - 2)
+    struct_mask[(*leading, ys + span_start, xs + span_start)] = 0
```
Comment (melisande-c) [inline src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py:458]: I was confused by adding `span_start` to `ys` and `xs` at first, it is not clear that `disp_mask_2d` is smaller than subpatch size. 

I think `_build_struct_pattern` and `_create_struct_mask` could probably be one function.

File: src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py [hunk 6]
```diff
@@ -400,16 +442,18 @@ def _create_struct_mask(
...
+    struct_mask[(*leading, center_idx, center_idx)] = 0  # center pixel
```
Comment (melisande-c) [inline src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py:459]: I think the centre pixel is covered by the struct mask already

File: tests/compat/transforms/test_old_pixel_manipulation.py [hunk 7]
```diff
@@ -16,13 +16,9 @@
...
-    _create_struct_mask,
-    _get_stratified_coords_torch,
+    _create_neg_center_pixel_mask,
+    _create_neg_struct_mask,
     _get_subpatch_coords,
-    median_manipulate,
-    uniform_manipulate,
 )
 from careamics.lightning.modules.n2v_utils.struct_mask_parameters import (
     StructMaskParameters,
```
Comment (melisande-c) [inline tests/compat/transforms/test_old_pixel_manipulation.py]: hmm I think these are meant to be imported from `compat.transforms`

File: tests/lightning/n2v_utils/test_pixel_manipulation.py
Comment (melisande-c) [inline tests/lightning/n2v_utils/test_pixel_manipulation.py:1]: Why the rename? If it's because of pytest name clashes you can add a `__init__.py` file to the directory to fix it I think.

File: tests/compat/transforms/test_old_pixel_manipulation.py [hunk 8]
```diff
@@ -588,7 +351,9 @@ def test_get_subpatch_coords(
 @pytest.mark.parametrize("n_dims", [2, 3])
 @pytest.mark.parametrize("subpatch_size", [5, 7, 11])
 def test_create_center_pixel_mask(n_dims: int, subpatch_size: int):
-    mask_tensor = _create_center_pixel_mask(n_dims, subpatch_size, torch.device("cpu"))
+    mask_tensor = _create_neg_center_pixel_mask(
```
Comment (melisande-c) [inline tests/compat/transforms/test_old_pixel_manipulation.py]: Since this is the current function and not compat it is now duplicated in `tests/lightning/n2v_utils/test_pixel_manipulation_.py`

However this version also parametrises subpatch size not just n_dims.

Same for tests below.

**Review by melisande-c — CHANGES_REQUESTED** [review]

File: src/careamics/lightning/modules/n2v_utils/struct_mask_parameters.py (hunk 1 above)
Comment (jdeschamps) [reply src/careamics/lightning/modules/n2v_utils/struct_mask_parameters.py]: Ok, that would make more sense indeed.

File: src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py (hunk 3 above)
Comment (jdeschamps) [reply src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py]: Ahah in my mind is it negative because my reference is the center pixel and the structN2V mask, as opposed to the new utility that creates a positive structN2V mask. I needed to mark the difference between the two.

I'll give it more thoughts.

File: src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py (hunk 4 above)
Comment (jdeschamps) [reply src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py:82]: What is the suggestion? The variable names?

Actually, I wanted to change "disp_mask_2d", that is not what it is (not a displacement matrix). Tbf in this PR I may have altered some variable names just because it was still a bit confusing to me.

File: src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py (hunk 5 above)
Comment (jdeschamps) [reply src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py:458]: I split `_build_struct_pattern` off because I wanted it to be reused in the negative structN2V mask creation used to exclude the mask from the median calculation.

The point being that the day we want a 2D arbitrary mask, we can just change that one `_build_struct_pattern` function and everything should work.

File: src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py (hunk 6 above)
Comment (jdeschamps) [reply src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py:459]: Ah true!

File: tests/lightning/n2v_utils/test_pixel_manipulation.py
Comment (jdeschamps) [reply tests/lightning/n2v_utils/test_pixel_manipulation.py:1]: Ah I didn't know that!

File: tests/compat/transforms/test_old_pixel_manipulation.py (hunk 7 above)
Comment (jdeschamps) [reply tests/compat/transforms/test_old_pixel_manipulation.py]: The imports were then wrong before as well then, will fix that.

File: tests/compat/transforms/test_old_pixel_manipulation.py (hunk 8 above)
Comment (jdeschamps) [reply tests/compat/transforms/test_old_pixel_manipulation.py]: I will have a look and clean that. Probably the imports were wrong from the start.

File: src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py (hunk 4 above)
Comment (melisande-c) [reply src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py:82]: Tbh not sure what my suggestion is, this whole thing is still pretty confusing, we can just leave it.

File: tests/compat/transforms/test_old_pixel_manipulation.py (hunk 7 above)
Comment (melisande-c) [reply tests/compat/transforms/test_old_pixel_manipulation.py]: yeah I think it must have got mixed up with the restructuring

File: src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py (hunk 5 above)
Comment (melisande-c) [reply src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py:458]: ok, makes sense

File: src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py (hunk 6 above)
Comment (jdeschamps) [reply src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py:459]: No actually, `_build_struct_pattern` creates a mask with the center being `0`, then in this particular function (which builds an opposite mask) it creates an array of `1`s and set to `0` every `1` from `_build_struct_pattern`.

So the center pixel set to 0 is necessary otherwise it will stay `1` and be included in the median manipulate.

Comment (jdeschamps) [reply src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py:459]: (yes these functions are nothing but opaque)

File: src/careamics/lightning/modules/n2v_utils/struct_mask_parameters.py (hunk 1 above)
Comment (jdeschamps) [reply src/careamics/lightning/modules/n2v_utils/struct_mask_parameters.py]: Done in https://github.com/CAREamics/careamics/pull/971/changes/8639f0360e8a0f8c44ec1b227f71d37e3a5f863d

File: src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py (hunk 3 above)
Comment (jdeschamps) [reply src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py]: In https://github.com/CAREamics/careamics/pull/971/changes/bff9aee187d7dcaf694394318fb1963ea4cb1022 I renamed the functions to `*exclusion_mask` for clarity.

File: src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py (hunk 4 above)
Comment (jdeschamps) [reply src/careamics/lightning/modules/n2v_utils/pixel_manipulation.py:82]: Renamed that particular variable in https://github.com/CAREamics/careamics/pull/971/commits/b261808026c47d17340a59f23fcf6332b8b4644d nonetheless.

File: tests/compat/transforms/test_old_pixel_manipulation.py (hunk 7 above)
Comment (jdeschamps) [reply tests/compat/transforms/test_old_pixel_manipulation.py]: This branch is now based on #976, which removes all `compat` stuff.

File: tests/compat/transforms/test_old_pixel_manipulation.py (hunk 8 above)
Comment (jdeschamps) [reply tests/compat/transforms/test_old_pixel_manipulation.py]: Same as above, `compat` removed by rebasing to #976

**Comment by jdeschamps** [comment]
`N2VManipulate` was too intertwined with `careamics.compat`, making it difficult to not add more technical debt. This PR is now based on #976, which gets rid of `compat` modules.

The changes are the following:
- structN2V parameters are now folded into a Pydantic model of their own (in `n2v_manipulate_config.py`), and it is used directly by the pixel manipulation utilities (`pixel_manipulation.py`)
- `Configuration` version now bumped to `0.3.0` because it is no longer compatible.
- Breaking change: the convenience functions parameter for the structN2V axis is now `struct_n2v_axes` (plural), to be coherent with the new configuration
- Few renaming here and there in `pixel_manipulation.py).

File: tests/lightning/n2v_utils/test_pixel_manipulation.py
Comment (jdeschamps) [reply tests/lightning/n2v_utils/test_pixel_manipulation.py:1]: Done in https://github.com/CAREamics/careamics/pull/971/changes/9de0896fff0d649ec2f0c77a891a1f84fc992bb8

File: tests/compat/transforms/test_old_pixel_manipulation.py (hunk 8 above)
Comment (melisande-c) [reply tests/compat/transforms/test_old_pixel_manipulation.py]: These original tests had more parametrisation for the `center_pixel_exclusion_mask` and the `struct_pixel_exclusion_mask`. There was also a test for even number subpatch sizes not being allowed

```
@pytest.mark.parametrize("n_dims", [2, 3])
@pytest.mark.parametrize("subpatch_size", [5, 7, 11])
```

File: tests/unit/lightning/modules/n2v_utils/test_pixel_manipulation.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        itertools.product(
+            [3, 4],  # BYX or BZYX
+            AXES,
+            SPAN,
+        )
+    ),
+)
+def test_create_neg_struct_mask(ndims, axes, span):
+    """Test that structN2V pattern is correctly excluded."""
+    subpatch_size = 11
```
Comment (melisande-c) [inline tests/unit/lightning/modules/n2v_utils/test_pixel_manipulation.py:86]: The previous tests parametrised subpatch_size and also tested that even sizes are not allowed. As mentioned in https://github.com/CAREamics/careamics/pull/971/changes/BASE..9b505ad6c2ec9239af26dc9fb3b5373ccfcf1e54#r3422173354

Comment (melisande-c) [reply tests/unit/lightning/modules/n2v_utils/test_pixel_manipulation.py:86]: Also the test name does not the function name anymore `neg -> exclusion`

File: src/careamics/config/algorithms/n2v_manipulation/n2v_manipulate_config.py [hunk 10]
```diff
@@ -1,12 +1,54 @@
...
+    ------
+    ValueError
+        If the value is even.
+    """
+    if v % 2 == 0:
+        raise ValueError("Size must be an odd number.")
+    return v
+
+
+class StructMaskParameters(BaseModel):
```
Comment (melisande-c) [inline src/careamics/config/algorithms/n2v_manipulation/n2v_manipulate_config.py]: Maybe `StructMaskConfig` would be more consistent.

**Review by melisande-c — APPROVED** [review]
Nice, makes much more sense now.

I left some comments on some small things, feel free to ignore.

File: tests/unit/lightning/modules/n2v_utils/test_pixel_manipulation.py (hunk 9 above)
Comment (melisande-c) [reply tests/unit/lightning/modules/n2v_utils/test_pixel_manipulation.py:86]: > The previous tests parametrised subpatch_size and also tested that even sizes are not allowed. As mentioned in https://github.com/CAREamics/careamics/pull/971/changes/BASE..9b505ad6c2ec9239af26dc9fb3b5373ccfcf1e54#r3422173354

Wait sorry, a version of these tests still exist in `tests/lightning/n2v_utils/test_pixel_manipulation.py`

File: tests/compat/transforms/test_old_pixel_manipulation.py (hunk 8 above)
Comment (melisande-c) [reply tests/compat/transforms/test_old_pixel_manipulation.py]: wait sorry a version of these tests still exist in `tests/lightning/n2v_utils/test_pixel_manipulation.py`

File: src/careamics/config/algorithms/n2v_manipulation/n2v_manipulate_config.py (hunk 10 above)
Comment (jdeschamps) [reply src/careamics/config/algorithms/n2v_manipulation/n2v_manipulate_config.py]: Oh yes, done in https://github.com/CAREamics/careamics/pull/971/changes/3209660cae5c5370bf16c802140a18bf692322d0

File: tests/unit/lightning/modules/n2v_utils/test_pixel_manipulation.py (hunk 9 above)
Comment (jdeschamps) [reply tests/unit/lightning/modules/n2v_utils/test_pixel_manipulation.py:86]: I did not see some "unit" tests in the old test file, I reimplemented them from scratch. In https://github.com/CAREamics/careamics/pull/971/changes/34462fc3dc72114f6a442b610a16e7ed411c5983 I moved the remaining previous tests into the correct test module.
